src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,166 @@
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = getNode(index);
+        T oldValue = node.data;
+        node.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        return removeNode(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+
+        for (Node<T> first = head; first != null; first = first.next) {
+            if (first.data == null) {
+                removeNode(first);
+                return true;
+            }
+            if (object != null && object.equals(first.data)) {
+                removeNode(first);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void insertLast(T data) {
+        Node<T> newNode = new Node<T>(tail, data, null);
+        if (tail != null) {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        if (head == null) {
+            head = newNode;
+        }
+        size++;
+    }
+
+    public void insertBefore(T element, Node<T> follow) {
+        Node<T> prior = follow.previous;
+        Node<T> newNode = new Node<>(prior, element, follow);
+        follow.previous = newNode;
+        if (prior == null) {
+            head = newNode;
+        } else {
+            prior.next = newNode;
+        }
+        size++;
+    }
+
+    private T removeNode(Node<T> removed) {
+        final T element = removed.data;
+        Node<T> next = removed.next;
+        Node<T> prev = removed.previous;
+
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+            removed.previous = null;
+        }
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.previous = prev;
+            removed.next = null;
+        }
+        removed.data = null;
+        size--;
+        return element;
+    }
+
+    private Node<T> getNode(int index) {
+        if (index < (size / 2)) {
+            Node<T> forward = head;
+            for (int k = 0; k < index; k++) {
+                forward = forward.next;
+            }
+            return forward;
+        } else {
+            Node<T> backward = tail;
+            for (int k = size - 1; k > index; k--) {
+                backward = backward.previous;
+            }
+            return backward;
+        }
+    }
+
+    private void checkBounds(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException(
+                    "This index: " + index + " is not valid for add operation");
+        }
+    }
+
+    private void checkElementIndex(int index) {
```

Comment:
these methods should have similar naming because they do practically the same, let's make code consistent
checkBounds(index)
checkAddBounds(index)

Reply:
done!